core/drivers/zynqmp_csu_aes.c

```diff
@@ -0,0 +1,426 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) Foundries Ltd. 2021
+ * Author: Jorge Ramirez <jorge@foundries.io>
+ */
+
+#include <config.h>
+#include <io.h>
+#include <kernel/boot.h>
+#include <kernel/delay.h>
+#include <kernel/dt.h>
+#include <libfdt.h>
+#include <malloc.h>
+#include <mm/core_memprot.h>
+#include <string.h>
+#include <tee/cache.h>
+
+#include "drivers/zynqmp_csudma.h"
+#include "drivers/zynqmp_csu_aes.h"
+#include "drivers/zynqmp_csu.h"
+
+/* CSU AES registers */
+#define AES_STS_OFFSET			0x00
+#define AES_KEY_SRC_OFFSET		0x04
+#define AES_KEY_LOAD_OFFSET		0x08
+#define AES_START_MSG_OFFSET		0x0C
+#define AES_RESET_OFFSET		0x10
+#define AES_KEY_CLR_OFFSET		0x14
+#define AES_CFG_OFFSET			0x18
+
+#define AES_KEY_LOAD			1
+#define AES_STS_AES_BUSY	        BIT(0)
+#define AES_STS_AES_KEY_ZEROED		BIT(8)
+#define AES_STS_KUP_ZEROED		BIT(9)
+#define AES_STS_KEY_INIT_DONE		BIT(4)
+#define AES_STS_GCM_TAG_OK		BIT(3)
+#define AES_START_MSG			1
+#define AES_CFG_ENC			1
+#define AES_CFG_DEC			0
+#define AES_RESET_SET			1
+#define AES_RESET_CLR			0
+#define AES_KEY_ZERO			BIT(0)
+#define AES_KUP_ZERO			BIT(1)
+
+#define AES_TIMEOUT_USEC		2000000
+
+enum aes_op { AES_DEC, AES_ENC };
+
+static TEE_Result aes_wait(uint32_t event, bool set)
+{
+	vaddr_t aes = core_mmu_get_va(ZYNQMP_CSU_AES_BASE, MEM_AREA_IO_SEC,
+				      ZYNQMP_CSU_AES_SIZE);
+	uint64_t tref = timeout_init_us(AES_TIMEOUT_USEC);
+	uint32_t status = 0;
+
+	if (!aes)
+		return TEE_ERROR_GENERIC;
+
+	while (!timeout_elapsed(tref)) {
+		status = io_read32(aes + AES_STS_OFFSET) & event;
+		if ((set && status == event) || (!set && status != event))
+			return TEE_SUCCESS;
+	}
+
+	return TEE_ERROR_GENERIC;
+}
+
+static TEE_Result aes_transfer_enc(const void *src, void *dst, size_t len,
+				   void *tag, const void *iv)
+{
+	uint8_t iv_padded[ZYNQMP_CSUDMA_MIN_SIZE]
+		__aligned(ZYNQMP_CSUDMA_ALIGN) = { 0 };
+	uint8_t dst_padded[SMALL_PAGE_SIZE]
+		__aligned(ZYNQMP_CSUDMA_ALIGN) = { 0 };
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (len + ZYNQMP_GCM_TAG_SIZE > SMALL_PAGE_SIZE) {
+		/* artificial but realistic limit for our use case */
+		EMSG("Input too large");
+		return TEE_ERROR_GENERIC;
+	}
+
+	ret = zynqmp_csudma_prepare();
+	if (ret) {
+		EMSG("DMA can't initialize");
+		return ret;
+	}
+
+	/* Prepare destination */
+	ret = zynqmp_csudma_transfer(ZYNQMP_CSUDMA_DST_CHANNEL,
+				     (void *)dst_padded,
+				     (len + ZYNQMP_GCM_TAG_SIZE) / 4, 0);
```

Comment:
`4` looks like a magic number.

Reply:
yeah, ok sizeof (uint32_t) then

Reply:
Could be worth mentioning this in the documentation of the function since `len` usually is number of bytes. Perhaps the parameter `len` in `zynqmp_csudma_transfer()` should be renamed to `word_count` or such?

Reply:
yes, you are right